Add absolute humidity component

DAVe3283 opened the pull request:
# What does this implement/fix?

Adds a component to calculate the absolute humidity provided air temperature (in °C) and relative humidity (in %) from other sensors. I use this with BME280 and BME680 sensors, but any should work.

I implemented 3 saturation vapor pressure equations: [Buck](https://en.wikipedia.org/wiki/Arden_Buck_equation), [Tetens](https://en.wikipedia.org/wiki/Tetens_equation), and [Wobus](https://wahiduddin.net/calc/density_altitude.htm). Defaults to Wobus, which seems to be the most accurate. However, most cheap humidity sensors are going to be, by far, the limitation here.

I believe I have worked out all the edge cases in this code, so bringing it upstream from <https://PigLab.ReaperLegion.net/home-automation/hass/esphome/custom-components/absolute-humidity>.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2723

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [X] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
sensor:
  - platform: absolute_humidity
    name: Absolute Humidity
    temperature: air_temperature
    humidity: relative_humidity

  # Use any temperature and relative humidity source, e.g. a BME280:
  - platform: bme280
    temperature:
      name: Temperature
      id: air_temperature
    pressure:
      name: Pressure
    humidity:
      name: Relative Humidity
      id: relative_humidity
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DAVe3283 committed:
Import Absolute Humidity component

https://PigLab.ReaperLegion.net/home-automation/hass/esphome/custom-components/absolute-humidity

DAVe3283 committed:
Fix terminology, add some docstrings

vidplace7 commented:
Working great here using the external_component-version (unable to test the PR-version for now since it introduces a new CONST)
```yaml
external_components:
  - source:
      type: git
      url: https://piglab.reaperlegion.net/home-automation/hass/esphome/custom-components/absolute-humidity.git
    components: [absolute_humidity]
```

Thanks a bunch for contributing this upstream @DAVe3283
My configs have been simplified greatly thanks to this (vs using template sensors)

jesserockz commented on esphome/components/absolute_humidity/sensor.py:
```diff
@@ -0,0 +1,60 @@
+        unit_of_measurement=UNIT_GRAMS_PER_CUBIC_METER,
+        icon=ICON_WATER,
+        accuracy_decimals=2,
+        device_class=DEVICE_CLASS_EMPTY,
```
This is the default and not needed here
```suggestion
```

jesserockz commented on esphome/components/absolute_humidity/sensor.py:
```diff
@@ -0,0 +1,60 @@
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await sensor.register_sensor(var, config)
```
```suggestion
    var = await sensor.new_sensor(config)
    await cg.register_component(var, config)
```

jesserockz commented on esphome/components/absolute_humidity/absolute_humidity.h:
```diff
@@ -0,0 +1,76 @@
+ public:
+  AbsoluteHumidityComponent() = default;
+
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { humidity_sensor_ = humidity_sensor; }
+  void set_equation(SaturationVaporPressureEquation equation) { equation_ = equation; }
```
```suggestion
  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { this->temperature_sensor_ = temperature_sensor; }
  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { this->humidity_sensor_ = humidity_sensor; }
  void set_equation(SaturationVaporPressureEquation equation) { this->equation_ = equation; }
```

jesserockz commented on esphome/components/absolute_humidity/absolute_humidity.h:
```diff
@@ -0,0 +1,76 @@
+
+  float temperature_{NAN};
+  float humidity_{NAN};
+  SaturationVaporPressureEquation equation_{WOBUS};
```
Always set by code generation here
```suggestion
  SaturationVaporPressureEquation equation_{};
```

jesserockz commented on esphome/components/absolute_humidity/absolute_humidity.cpp:
```diff
@@ -0,0 +1,183 @@
+      break;
+  }
+
+  ESP_LOGCONFIG(TAG, "Sources");
+  LOG_SENSOR("  ", "Temperature", this->temperature_sensor_);
+  LOG_SENSOR("  ", "Relative Humidity", this->humidity_sensor_);
```
Since these sensors will be dumped by their own code, I dont think its necessary to dump the full log here with LOG_SENSOR, can probably just log the name

jesserockz reviewed (changes requested):
Looks good, not gonna try understand the formulas...haha

One question, is `double` precision needed everywhere? We try to avoid `double `if we can in ESPHome and just use `float` for consistency. 
Its fine if it does make a difference...

DAVe3283 commented:
> Looks good, not gonna try understand the formulas...haha
> 
> One question, is `double` precision needed everywhere? We try to avoid `double `if we can in ESPHome and just use `float` for consistency. Its fine if it does make a difference...

It looks like it doesn't make a practical difference. It is different, but well beyond the default 2 decimal places we report (I only see a difference once you get 5-6 decimals in). I will change it to use float as I address your other feedback.

DAVe3283 committed:
Merge remote-tracking branch 'remotes/origin/dev' into absolute_humidity

DAVe3283 committed:
Switch from double to float

https://github.com/esphome/esphome/pull/4519#pullrequestreview-1327615169
The additional precision doesn't matter in practice.

DAVe3283 committed:
Address code review suggestions

DAVe3283 committed:
Lint code

jesserockz reviewed (approved):
Looks good, thanks for making the changes so quickly =)
